### PR TITLE
Add a `strategy` tactic

### PR DESCRIPTION
**Pull request by JasonGross**
Useful for guarding calls to `unfold` or `cbv` to ensure that, e.g.,
`Opaque foo` doesn't break some automation which tries to unfold `foo`.

**Kind:** feature

- [x] Added / updated test-suite
- [x] Corresponding documentation was added / updated (including any warning and error messages added / removed / modified).
- [x] Entry added in the changelog (see https://github.com/coq/coq/tree/master/doc/changelog#unreleased-changelog for details).

<strike>I'm confused about `gec_gen` vs `Entry.create` vs `make_gen_entry`, and also I'm getting
```
File "./theories/Init/Notations.v", line 128, characters 0-32:
Error: Anomaly "grammar function not found: strategy_level."
Please report at http://coq.inria.fr/bugs/.
```
and don't know how to fix it...</strike>

### CODE REVIEW

**Comment by ppedrot** [comment]
You should look at `Tacentries.argument_extend` and use a variant of this code to create a hand-written argument. Ssreflect also does this kind of hack, in a less elegant way.

**Comment by JasonGross** [comment]
@ppedrot Thanks, I think I've managed to get `strategy_level` working.  Now I'm stuck on `smart_global`.  In particular, I'm not sure how to write `interp_smart_global` for `tacinterp.ml` nor how to write `intern_smart_global` for `tacintern.ml`.

File: doc/changelog/04-tactics/11029-add-strategy.rst
Comment (Zimmi48) [inline doc/changelog/04-tactics/11029-add-strategy.rst]: Nobody is really sure what "vernacular" means, whereas everyone agrees on the definition of a "command", so better use the latter:

```suggestion
- New tactic :tacn:`strategy` added which behaves like the command
```

File: doc/sphinx/proof-engine/tactics.rst
Comment (Zimmi48) [inline doc/sphinx/proof-engine/tactics.rst]: Ditto

File: plugins/ltac/extratactics.mlg
Comment (ppedrot) [inline plugins/ltac/extratactics.mlg]: This should probably be put somewhere else, behind a mli.

**Review by ppedrot — CHANGES_REQUESTED** [review]
For code cleanliness purposes, I think you should use the `Tacentries.argument_extend` function. It lets you provide an arbitrary parsing entry, so this should be fine. This way, all the nasty declarations would be centralized in one point in the code.

File: doc/sphinx/proof-engine/tactics.rst
Comment (JasonGross) [reply doc/sphinx/proof-engine/tactics.rst]:
```suggestion
   command.
```

File: plugins/ltac/extratactics.mlg
Comment (JasonGross) [reply plugins/ltac/extratactics.mlg]: The only difference between this and `vernac_set_strategy` is `match r with` vs `match smart_global r with`.  What do you think of moving this function to vernacentries (and exposing it in the .mli), and then making `vernac_set_strategy` first `map smart_global` and then call this function?

**Comment by JasonGross** [comment]
@ppedrot I don't see how to do that without a lot of code duplication, or massive changes to the existing infrastructure.  As I understand it, the code from tacinterp and tacintern and tacsubst is not exposed to other files, but I make use of functions defined in those files for interp and intern and subst.  So I'm not sure what exactly it is you're suggesting.  Am I misunderstanding something?
Alternatively, if you want to demonstrate by updating existing code to use `Tacentries.argument_extend` for `ref`/`reference`, `int`/`integer`, and `int_or_var`, I will happily follow suite with `smart_global`, `strategy_level` and `strategy_level_or_var`.  (Those are the entries that I cargo-culted to figure out how to do these ones.)  Also feel free to push a commit to my branch implementing your suggestion.

**Comment by JasonGross** [comment]
Also @ppedrot, it seems like it would be poor form to use something in the Ltac plugin to register the Pcoq grammar entries for `smart_global` and `strategy_level`, given that these are both used outside of Ltac in the Strategy vernacular, which is part of the base vernacular AST.  (`strategy_level_or_var` seems fine in principle, but I don't have any examples of an `or_var` entry added via `argument_extend` to work from)

**Comment by JasonGross** [comment]
@coq/doc-maintainers I've pushed another commit documenting the additions to the tactic notation entries that I had previously forgotten to document.

**Review by Zimmi48 — APPROVED** [review]
I am not very knowledgeable of the notation chapter but this looks good to me.

File: plugins/ltac/extratactics.mlg
Comment (JasonGross) [reply plugins/ltac/extratactics.mlg]: Moved

**Comment by ppedrot** [comment]
@JasonGross I took the liberty to factorize the argument declaration in extraargs via a macro. This passes the test-suite so it should probably behave the same.

Also, maybe the grammar entry should actually be moved to `G_vernac`. It's not used below that, and it's quite a specific entry to be in the primitive module.

File: plugins/ltac/extratactics.mlg
Comment (ppedrot) [inline plugins/ltac/extratactics.mlg]: This is extremely fragile and likely a piece of code that is going to break when we move towards a more functional implementation. This function is actually no more than a wrapper around its `Global` counterpart, as the VM-acting function in `Vernacentries` is a huge bluff, it's just a no-op. Thus this should probably be defined by tweaking the tactic-local environment with `tclENV` instead, and pray that no one is using the imperative global environment in the tactics you're interested in.

Comment (JasonGross) [reply plugins/ltac/extratactics.mlg]: Does `tclENV` play nicely with `abstract`?  Note the following in the test-suite:
```coq
(* test that the strategy tactic does persist through abstract *)
Opaque id.
Goal id 0 = 0.
  strategy expand [id].
  assert (id (fact 100) = fact 100) by abstract reflexivity.
  reflexivity.
Defined.
```

> This function is actually no more than a wrapper around its `Global` counterpart, as the VM-acting function in `Vernacentries` is a huge bluff, it's just a no-op.

What is the "VM-acting function" you refer to?  Is it the `Csymtable.set_opaque_const` / `Csymtable.set_transparent_const` call?  Or the call to `Lib.add_anonymous_leaf`?

Are you suggesting we instead do something like
```ocaml
  let glob_key r =
    match r with
      | GlobRef.ConstRef sp -> ConstKey sp
      | GlobRef.VarRef id -> VarKey id
      | _ -> user_err Pp.(str
          "cannot set an inductive type or a constructor as transparent") in
  tclENV >>= fun env ->
  let env = List.fold_left (fun env q -> Environ.set_oracle env (Conv_oracle.set_strategy (Environ.oracle env) (glob_key q) v)) env ql in
  tclSETENV env
```
?  (And if so, which `.ml` file should this live in?)

**Comment by JasonGross** [comment]
> I took the liberty to factorize the argument declaration in extraargs via a macro. This passes the test-suite so it should probably behave the same.

@ppedrot Thanks!  Is `smart_global` good as it stands, or should it also be moved?

**Comment by JasonGross** [comment]
> Also, maybe the grammar entry should actually be moved to `G_vernac`. It's not used below that, and it's quite a specific entry to be in the primitive module.

I have no opinion on this; feel free to push a commit doing so (or ask me to make a particular change) if you'd like.

File: plugins/ltac/extratactics.mlg
Comment (JasonGross) [reply plugins/ltac/extratactics.mlg]: Does `tclENV` play nicely with `abstract`?  Note the following in the test-suite:
```coq
(* test that the strategy tactic does persist through abstract *)
Opaque id.
Goal id 0 = 0.
  strategy expand [id].
  assert (id (fact 100) = fact 100) by abstract reflexivity.
  reflexivity.
Defined.
```

> This function is actually no more than a wrapper around its `Global` counterpart, as the VM-acting function in `Vernacentries` is a huge bluff, it's just a no-op.

What is the "VM-acting function" you refer to?  Is it the `Csymtable.set_opaque_const` / `Csymtable.set_transparent_const` call?  Or the call to `Lib.add_anonymous_leaf`?

Are you suggesting we instead do something like
```ocaml
  let glob_key r =
    match r with
      | GlobRef.ConstRef sp -> ConstKey sp
      | GlobRef.VarRef id -> VarKey id
      | _ -> user_err Pp.(str
          "cannot set an inductive type or a constructor as transparent") in
  tclENV >>= fun env ->
  let env = List.fold_left (fun env q -> Environ.set_oracle env (Conv_oracle.set_strategy (Environ.oracle env) (glob_key q) v)) env ql in
  tclSETENV env
```
?  (And if so, which `.ml` file should this live in?)

Comment (ppedrot) [reply plugins/ltac/extratactics.mlg]: > Does tclENV play nicely with abstract?

Unclear even to me, but worth testing.

> What is the "VM-acting function" you refer to?

The Csymtable. I did not realize that this function was calling `add_leaf` which makes it even worse. No tactic should mess with the libobject stack, the only one I thought was doing that was precisely abstract.

> Are you suggesting we instead do something like

If this works with abstract, then yes. Otherwise, just use the `Global.set_strategy` function, or even better maybe both this and the `tclENV` solution to make the code future proof.

Comment (JasonGross) [reply plugins/ltac/extratactics.mlg]: @ppedrot I tried this:
```ocaml
{

let set_strategy lvl ql =
  let glob_key r =
    match r with
    | GlobRef.ConstRef sp -> ConstKey sp
    | GlobRef.VarRef id -> VarKey id
    | _ -> user_err Pp.(str
        "cannot set an inductive type or a constructor as transparent") in
  Proofview.tclENV >>= fun env ->
  let env = List.fold_left (fun env q ->
    Environ.set_oracle env
      (Conv_oracle.set_strategy (Environ.oracle env) (glob_key q) lvl)) env ql in
  Proofview.Unsafe.tclSETENV env

}

TACTIC EXTEND strategy
| [ "strategy" strategy_level_or_var(v) "[" ne_smart_global_list(q) "]"  ] ->
  { set_strategy v q }
END
```
and it does not work so well.  It works for `strategy transparent [id]; unfold id`, but fails for `strategy transparent [id]. unfold id.`  How do I set the tactic environment in a way that persists across periods?

Comment (JasonGross) [reply plugins/ltac/extratactics.mlg]: @ppedrot I've updated the code to call `Global.set_strategy` more directly, which does seem to work fine.

Comment (ejgallego) [reply plugins/ltac/extratactics.mlg]: @JasonGross unfortunately in the current design you cannot update the global env from tactic code, this will create problem for example with workers. If you need to set the strategy in tactics that bit of state will have to be moved to the proof state.

Comment (JasonGross) [reply plugins/ltac/extratactics.mlg]: @ejgallego Sure, I noticed that if I do `Opaque id. par: strategy transparent [id]; shelve. Unshelve. unfold id.`, that fails. (I don't think there's any other way to get two bits of the same proof to be spread across workers, right?)  You're saying it's bad to commit this tactic when things are in such a state, because it creates technical debt, even if the tactic is already useful even if it doesn't work well with `par:`?  (I could insert a notice in the manual that this tactic doesn't play well with `par:`, for the time being.)

Indeed, setting the strategy in tactics is quite important (and eventually I want a way to make it persist long enough that it is still in effect during the running of `Qed`), so I think this should be moved to the proof state.

Comment (ejgallego) [reply plugins/ltac/extratactics.mlg]: I'd say let's move it to the proof state now [if the rest of the devs agree], IMVHO we have enough "corner cases" in Coq as to create more inconsistencies.

Specially when we the KNOWN_ISSUES file was rejected so we don't even document them.

Comment (JasonGross) [reply plugins/ltac/extratactics.mlg]: > I'd say let's move it to the proof state now

Sounds good.  Is this a PR you're willing to make, or a change I should try to make (and if me, should it be in this PR, or a separate one)?

Comment (ejgallego) [reply plugins/ltac/extratactics.mlg]: We can make it in a different PR, let's hear from @ppedrot tho first.

Comment (ppedrot) [reply plugins/ltac/extratactics.mlg]: I am not sure I understand the plan. The environment is already part of the proof state, and it contains the oracle state as well. What do you want to add?

Comment (ejgallego) [reply plugins/ltac/extratactics.mlg]: Not really, the environ that is part of the proof state is not the `Global` env, so effectively you have two environs; as pointed out here unless you call `Global.set_oracle` the changes won't be preserved for example across `run_tactic`.

**Review by maximedenes — APPROVED** [review]
`conv_oracle` changes are ok

File: plugins/ltac/extratactics.mlg
Comment (ejgallego) [inline plugins/ltac/extratactics.mlg]: Global modification of the environment should not be allowed from tactic code.

**Review by ejgallego — CHANGES_REQUESTED** [review]

File: plugins/ltac/extratactics.mlg
Comment (JasonGross) [reply plugins/ltac/extratactics.mlg]: Is the solution to move this part of the global state to proof state?  (I tried figuring out how to do that but couldn't make heads or tails of what's in proof state and what's not.)

Comment (JasonGross) [reply plugins/ltac/extratactics.mlg]: @ejgallego Ping

Comment (ejgallego) [reply plugins/ltac/extratactics.mlg]: Proof state is what is on the `engine/proofview` monad, we had some discussion about it in the Coq Call.

In order to add the state you would indeed augment the state monad there, but IMHO that's a risky proposition and that could have some performance impact; I dunno, the thing doesn't seem to be designed as to be extensible. Relevant code in `proofview_monad.ml`:
```
module P = struct

  type s = proofview * Environ.env

```
see also `logic_monad.ml`

Comment (SkySkimmer) [reply plugins/ltac/extratactics.mlg]: The strategy stuff is already in the env, why do you say to change that part of the code?

Comment (ejgallego) [reply plugins/ltac/extratactics.mlg]: To provide a bit more of context, what we are trying to do for now is to have the proof state not to depend on `Global`, you can see that the state may not be synchronized properly at some points, for example w.r.t. environmnets; I need to sync with @ppedrot to see what the steps are [I have some patches sitting on my HD], but in short; we'd like to finish pushing up `Global` up before we move towards making the proof state easier to tweak [in a sense the more things we put inside the larger the risk for problems later when we attempt the move]

Comment (ejgallego) [reply plugins/ltac/extratactics.mlg]: > The strategy stuff is already in the env, why do you say to change that part of the code?

Oh sorry, indeed you had mentioned that in the call; so it may be doable without modifying the state indeed.

Comment (SkySkimmer) [reply plugins/ltac/extratactics.mlg]: >Oh sorry, indeed you had mentioned that in the call; so it may be doable without modifying the state indeed.

The problem is that @JasonGross wants this to work
~~~coq
strategy bla.
foo_tac. (* use the new strategy *)
~~~
but the env in the monad gets reset after every `.` using the global env.

Comment (ejgallego) [reply plugins/ltac/extratactics.mlg]: Indeed, thanks for the reminder @SkySkimmer ; holidays have done a bit of damage to my brain. I guess that's the reason I mentioned the cleanups, I have a patch in my HD that should help with this problem, as it tries to improve how the envs are handled, I guess I'll try to finish it soon.

**Comment by ppedrot** [comment]
What do we do with this then?

**Comment by ejgallego** [comment]
> What do we do with this then?

So far seems blocked on the two directions:

- If we allow `Global.set_strategy` inside tactics it both creates problems with workers and with future restrictions on the API.
- Setting it functionally is buggy as for example the strategy doesn't persist after the `.` , due to the wrong copy of the env being modified; this requires non-trivial cleanup in the engine side I think.

**Comment by ppedrot** [comment]
I don't understand why the locally-restricted version is bad (i.e. the one that does not survive a dot). Most options could be passed that way without interfering with the document manager.

**Comment by ejgallego** [comment]
> I don't understand why the locally-restricted version is bad (i.e. the one that does not survive a dot). Most options could be passed that way without interfering with the document manager.

I'm not sure I understand what you mean; when you have `tac1. tac2`, `tac2` anyways depends on the `evar_map` that `tac1` outputs so the document manager can do little here.

**Comment by ppedrot** [comment]
My point is: if you want to affect the global state, use a vernacular command. Setting a value inside a tactic should respect the scope instead, à la dynamic binding.

**Comment by maximedenes** [comment]
Yeah, I was also wondering why a `with_strategy` tactical was not enough.

**Comment by ejgallego** [comment]
> Setting a value inside a tactic should respect the scope instead, à la dynamic binding.

Well that is true for some parts of the env, not for the others.

**Comment by JasonGross** [comment]
> Yeah, I was also wondering why a `with_strategy` tactical was not enough.

I can indeed turn this into a `with_strategy` tactic, which would already be useful, but I have two issues with this:
1. It forces me to write all my tactics in CPS (indeed the wider use-case is guarding a call to `unfold`, for which `with_strategy` is fine, but the use case I have in rewriters / fiat-crypto is in setting the strategy of constants defined via `transparent_abstract`, which must be persisted across the entire rest of the proof)
2. (and more importantly, for my use-case) it does not persist into `Defined`, and so if I set the strategy and then do some reduction that is reflected in a conversion problem in the proof term, the resulting `Defined` can blow up.

**Comment by JasonGross** [comment]
Note to self: the current version doesn't work well enough, because effects show up *only* after the `.`.  So I think I need a combination of `tclSETENV` and `Global.set_strategy`, as is used in #12129

**Comment by ppedrot** [comment]
@JasonGross isn't this superseded by #12129?

**Comment by JasonGross** [comment]
Not quite.  I still have the use case of wanting to script the setting of strategies for tactics defined by `transparent_abstract`, and having these strategies persist globally past `Defined`.  But it seems this will have to wait for a reworking of `abstract` and how constants are handled, so let me make this a draft PR, since indeed the other use cases are handled by the other PR

**Comment by JasonGross** [comment]
Btw, I found a new reason why `with_strategy` isn't sufficient: without tweaking the libobject stack, `coqchk` will fall over on some examples (see #12200 for an example where even the libobject stack isn't enough to prevent `coqchk` from falling over (performance-wise) on valid .vo files).

**Comment by ppedrot** [comment]
The more I struggle with kernel side-effects and similar global mutable state, the less I believe this PR should make it. Rather than introducing more state desynchronization, I think we should first sort out the general mess and implement this when we have a clear API in mind for this. Unluckily it's a fairly pervasive change that touches many components, so not easy. In the meantime I don't think there is much point in keeping this open since it will require a full rewrite anyways. Thus, closing.

**Pull request closed without merging**
